File: download/src/main/java/com/bbou/download/BaseDownloadFragment.java
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>.
 */

package com.bbou.download;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.Log;

import java.util.UUID;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.work.Data;
import androidx.work.WorkInfo;

/**
 * Base download fragment
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
abstract public class BaseDownloadFragment extends AbstractDownloadFragment
{
	/**
	 * Log tag
	 */
	static protected final String TAG = "BaseDownloadF";

	// L I F E C Y C L E

	@Override
	public void onCreate(final Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);

		// arguments
		unmarshal();

		// main receiver
		Log.d(TAG, "Register main receiver");

		final IntentFilter intentFilter = new IntentFilter();
		intentFilter.addAction(ACTION_DOWNLOAD_CANCEL);
		intentFilter.addCategory(Intent.CATEGORY_DEFAULT);

		ContextCompat.registerReceiver(requireContext(), this.cancelReceiver, intentFilter, ContextCompat.RECEIVER_NOT_EXPORTED);

		// notifications
		Notifier.initChannels(requireContext());
	}

	@Override
	public void onDestroy()
	{
		super.onDestroy();

		// main receiver
		Log.d(TAG, "Unregister main receiver");
		requireContext().unregisterReceiver(this.cancelReceiver);
	}

	// A R G U M E N T S

	abstract protected void unmarshal();

	// S T A T U S

	/**
	 * Result
	 */
	@Nullable
	protected Boolean success;

	/**
	 * Cancel pending
	 */
	protected boolean cancel;

	/**
	 * Cause
	 */
	@Nullable
	protected String cause;

	/**
	 * Exception
	 */
	@Nullable
	protected String exception;

	/**
	 * Downloading flag (prevents re-entrance)
	 */
	protected boolean downloading = false;

	/**
	 * Downloaded progress when status was read
	 */
	long progressDownloaded = 0;

	/**
	 * Total progress when status was read
	 */
	long progressTotal = 0;

	// B R O A D C A S T

	/**
	 * New data available request runnable, run when new data becomes available
	 */
	protected Runnable requestNew;

	/**
	 * Kill old data request runnable, run when old data should be discarded
	 */
	protected Runnable requestKill;

	/**
	 * Set new data available request runnable
	 *
	 * @param requestNew new data available request runnable
	 */
	public void setRequestNew(final Runnable requestNew)
	{
		this.requestNew = requestNew;
	}

	/**
	 * Set kill old data request runnable
	 *
	 * @param requestKill new data available request runnable
	 */
	public void setRequestKill(final Runnable requestKill)
	{
		this.requestKill = requestKill;
	}

	// N O T I F I C A T I O N

	/**
	 * Notification id key
	 */
	static public final String NOTIFICATION_ID = "notification_id";

	/**
	 * Id for the current notification
	 */
	static protected int notificationId = 0;

	/**
	 * Fire UI notification
	 *
	 * @param context        context
	 * @param notificationId notification id
	 * @param type           notification
	 * @param args           arguments
	 */
	abstract protected void fireNotification(@NonNull final Context context, int notificationId, @NonNull final Notifier.NotificationType type, final Object... args);

	// O B S E R V E R

	protected final androidx.lifecycle.Observer<WorkInfo> observer = (wi) -> {

		// progress
		if (this.downloading) // drop event if not
		{
			Data data = wi.getProgress();
			progressDownloaded = data.getLong(DownloadWork.PROGRESS, 0);
			progressTotal = data.getLong(DownloadWork.TOTAL, 0);
			float progress = (float) progressDownloaded / progressTotal;
			fireNotification(requireContext(), notificationId, Notifier.NotificationType.UPDATE, progress);
			Log.d(TAG, "Observed progress " + progressDownloaded + '/' + progressTotal);
		}

		// state
		WorkInfo.State state = wi.getState();
		switch (state)
		{
			// the WorkRequest is enqueued and eligible to run when its Constraints are met and resources are available
			case ENQUEUED:
				Log.d(TAG, "Observed enqueued");
				break;

			// the WorkRequest is currently being executed.
			case RUNNING:
				Log.d(TAG, "Observed running");
				this.downloading = true; // confirm
				fireNotification(requireContext(), notificationId, Notifier.NotificationType.START);
				break;

			// the WorkRequest has completed in a successful state.
			case SUCCEEDED:
				Log.d(TAG, "Observed succeeded");

				// state
				this.downloading = false;
				this.success = true;
				this.exception = null;
				this.cause = null;

				// record
				final Data data = wi.getOutputData();
				final String fromUrl = data.getString(DownloadWork.ARG_FROM);
				final long date = data.getLong(DownloadWork.DATE, -1);
				final long size = data.getLong(DownloadWork.SIZE, -1);
				final String etag = data.getString(DownloadWork.ETAG);
				final String version = data.getString(DownloadWork.VERSION);
				final String staticVersion = data.getString(DownloadWork.STATIC_VERSION);
				Settings.recordDatapackSource(requireContext(), fromUrl, date, size, etag, version, staticVersion);

				// fire notification
				fireNotification(requireContext(), notificationId, Notifier.NotificationType.FINISH, true);

				// fire onDone
				onDone(Status.STATUS_SUCCEEDED);
				break;

			// the WorkRequest has completed in a failure state.
			// All dependent work will also be marked as {@code #FAILED} and will never run.
			case FAILED:
				Log.d(TAG, "Observed failed");

				// state
				this.downloading = false; // release
				final Data errData = wi.getOutputData();
				this.success = false;
				this.exception = errData.getString(DownloadWork.EXCEPTION);
				this.cause = errData.getString(DownloadWork.EXCEPTION_CAUSE);

				// fire notification
				fireNotification(requireContext(), notificationId, Notifier.NotificationType.FINISH, false);

				// fire onDone
				onDone(Status.STATUS_FAILED);
				break;

			// the WorkRequest is currently blocked because its prerequisites haven't finished successfully.
			case BLOCKED:
				Log.d(TAG, "Observed block");
				break;

			// the WorkRequest has been cancelled and will not execute.
			// All dependent work will also be marked as #CANCELLED and will not run.
			case CANCELLED:
				Log.d(TAG, "Observed cancel");

				// state
				this.downloading = false;
				this.success = false;
				this.cancel = true;
				this.exception = null;
				this.cause = null;

				// fire notification
				fireNotification(requireContext(), notificationId, Notifier.NotificationType.CANCEL, false);

				// fire onDone
				onDone(Status.STATUS_CANCELLED);
				break;
		}
	};

	// C A N C E L   B R O A D C A S T / R E C E I V E R

	public static final String ACTION_DOWNLOAD_CANCEL = "action_cancel_download";

	/**
	 * Broadcast receiver for cancel events
	 */
	@NonNull
	protected final BroadcastReceiver cancelReceiver = new BroadcastReceiver()
	{
		@Override
		public void onReceive(final Context context, @NonNull final Intent intent)
		{

			Log.d(TAG, "Received broadcast request");
			String action = intent.getAction();
			if (ACTION_DOWNLOAD_CANCEL.equals(action))
			{
				Log.d(TAG, "Received cancel broadcast request");

				// do
				cancel();
			}
		}
	};

	public static Intent makeCancelIntent(@NonNull final Context context, int notificationId)
	{
		final Intent intent = new Intent();
		intent.setPackage(context.getApplicationContext().getPackageName());
		intent.setAction(ACTION_DOWNLOAD_CANCEL);
		intent.addCategory(Intent.CATEGORY_DEFAULT);
		intent.putExtra(NOTIFICATION_ID, notificationId);
		return intent;
	}

	public static PendingIntent makePendingIntent(@NonNull final Context context, @NonNull final Intent intent)
	{
		final int uid = (int) System.currentTimeMillis();
		final int flags = android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M ? //
				PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_CANCEL_CURRENT : //
				PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_CANCEL_CURRENT;
		return PendingIntent.getBroadcast(context, uid, intent, flags);
	}

	public static void requestCancel(@NonNull final Context context, int notificationId)
	{
		Intent intent = makeCancelIntent(context, notificationId);
		Log.d(TAG, "Sending cancel request (broadcast intent)");
		context.sendBroadcast(intent);
	}

	public static void requestPendingCancel(@NonNull final Context context, int notificationId)
	{
		try
		{
			Intent intent = makeCancelIntent(context, notificationId);
			PendingIntent pendingIntent = makePendingIntent(context, intent);
			Log.d(TAG, "Pending intent " + pendingIntent);
			Log.d(TAG, "Pending intent creator package " + pendingIntent.getCreatorPackage());
			Log.d(TAG, "Sending cancel request (sent pending intent)");
			// pendingIntent.send();
			pendingIntent.send(1313, (pendingIntent1, intent1, resultCode, resultData, resultExtras) -> Log.d(TAG, "Sent pending intent " + resultCode), null);
		}
		catch (PendingIntent.CanceledException e)
		{
			throw new RuntimeException(e);
		}
	}

	// C O N T R O L

	/**
	 * Work uuid
	 */
	@Nullable
	protected UUID uuid;

	/**
	 * Cancel download
	 */
	protected void cancel()
	{
		this.cancel = true;
		if (this.uuid != null)
		{
			DownloadWork.stopWork(requireContext(), uuid);
		}
	}

	// S T A T U S

	/**
	 * Download status translation to the abstract layer
	 *
	 * @param progress container to return progress as well
	 */
	@Override
	synchronized protected Status getStatus(@Nullable final Progress progress)
	{
		// progress
		if (progress != null)
		{
			progress.downloaded = this.progressDownloaded;
			progress.total = this.progressTotal;
		}

		// status
		if (this.cancel)
		{
			return Status.STATUS_CANCELLED;
		}
		else if (this.downloading)
		{
			return Status.STATUS_RUNNING;
		}
		else
		{
			if (this.success == null)
			{
				return Status.STATUS_PENDING;
			}
			else
			{
				return this.exception == null && this.success ? Status.STATUS_SUCCEEDED : Status.STATUS_FAILED;
			}
		}
	}

	/**
	 * Download reason
	 */
	@Nullable
	@Override
	protected String getReason()
	{
		if (this.exception != null)
		{
			return this.exception;
		}
		if (this.cause != null)
		{
			return this.cause;
		}
		return null;
	}

	@Override
	protected void show()
	{
	}
}

File: download_common/src/main/java/com/bbou/download/BaseDownloadFragment.kt
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>.
 */
package com.bbou.download

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.CallSuper
import androidx.annotation.UiThread
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import androidx.fragment.app.Fragment
import com.bbou.download.Keys.CANCEL_BTN_STATE
import com.bbou.download.Keys.DEPLOY_BTN_STATE
import com.bbou.download.Keys.DOWNLOAD_BTN_BACKGROUND_STATE
import com.bbou.download.Keys.DOWNLOAD_BTN_IMAGE_STATE
import com.bbou.download.Keys.DOWNLOAD_BTN_STATE
import com.bbou.download.Keys.DOWNLOAD_FROM_ARG
import com.bbou.download.Keys.DOWNLOAD_TARGET_FILE_ARG
import com.bbou.download.Keys.MD5_BTN_STATE
import com.bbou.download.Keys.PROGRESS_STATE
import com.bbou.download.Keys.PROGRESS_STATUS_STATE
import com.bbou.download.Keys.RENAME_FROM_ARG
import com.bbou.download.Keys.RENAME_TO_ARG
import com.bbou.download.Notifier.Companion.ACTION_DOWNLOAD_CANCEL
import com.bbou.download.common.R
import com.bbou.download.preference.Settings
import com.bbou.download.storage.FormatUtils.formatAsInformationString
import java.io.File
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.properties.Delegates

/**
 * Base download fragment that handles
 * - common arguments
 * - status including progress
 * - UI components
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
abstract class BaseDownloadFragment : Fragment() {

    // C O N T E X T

    protected lateinit var appContext: Context

    // S T A T U S

    /**
     * Whether one download is in progress
     */
    protected var isDownloading = AtomicBoolean(false)

    /**
     * Status, describes download status
     */
    enum class Status(
        /**
         * String resource id
         */
        val res: Int,
    ) {

        /**
         * Pending
         */
        STATUS_PENDING(R.string.status_download_pending),

        /**
         * Running
         */
        STATUS_RUNNING(R.string.status_download_running),

        /**
         * Paused
         */
        STATUS_PAUSED(R.string.status_download_paused),

        /**
         * Succeeded
         */
        STATUS_SUCCEEDED(R.string.status_download_successful),

        /**
         * Failed
         */
        STATUS_FAILED(R.string.status_download_fail),

        /**
         * Cancelled
         */
        STATUS_CANCELLED(R.string.status_download_cancelled);

        /**
         * Make status string
         *
         * @param context context
         * @return status string
         */
        fun toString(context: Context): String {
            return context.getString(res)
        }
    }

    /**
     * Id for the current notification
     */
    var notificationId = Notifier.notificationId()

    // I N  D A T A

    /**
     * Status
     */
    protected var status: Status? = null

    /**
     * Progress
     */
    protected var progress: Pair<Long, Long>? = 0L to 0L

    // A R G S

    /**
     * Download uri
     */
    protected var downloadUrl: String? = null

    /**
     * Rename source
     */
    protected var renameFrom: String? = null

    /**
     * Rename destination
     */
    protected var renameTo: String? = null

    /**
     * Mode
     */
    protected var mode: Settings.Mode? = null

    /**
     * Unmarshal arguments
     */
    protected open fun unmarshal() {

        // mode
        val modeStr = arguments?.getString(Keys.DOWNLOAD_MODE_ARG)
        mode = modeStr?.let { Settings.Mode.valueOf(modeStr) }

        // download source arg
        downloadUrl = arguments?.getString(DOWNLOAD_FROM_ARG)
        if (downloadUrl == null || downloadUrl!!.isEmpty()) {
            val message = appContext.getString(R.string.status_download_error_null_download_url)
            warn(message)
        }

        // rename
        renameFrom = arguments?.getString(RENAME_FROM_ARG)
        renameTo = arguments?.getString(RENAME_TO_ARG)
    }

    // W I D G E T S

    /**
     * Progress bar
     */
    private lateinit var progressBar: ProgressBar

    /**
     * Progress status view
     */
    private lateinit var progressStatus: TextView

    /**
     * Status view
     */
    private lateinit var statusTextView: TextView

    /**
     * Cancel downloads button
     */
    private lateinit var cancelButton: Button

    /**
     * MD5 button
     */
    protected var deployButton: Button? = null

    /**
     * MD5 button
     */
    protected var md5Button: Button? = null

    /**
     * Download button
     */
    private lateinit var downloadButton: ImageButton

    // R E S O U R C E S

    /**
     * Layout id
     */
    abstract val layoutId: Int

    /**
     * Saved resource id of image applied to download button
     */
    private var downloadButtonImageResId by Delegates.notNull<Int>()

    /**
     * Saved resource id of background applied to download button
     */
    private var downloadButtonBackgroundResId by Delegates.notNull<Int>()

    // L I F E C Y C L E

    /**
     * onCreate
     *
     * @param savedInstanceState saved instance state
     */
    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        // store app context
        appContext = requireContext().applicationContext

        // unmarshal arguments
        unmarshal()

        // cancel receiver
        Log.d(TAG, "Register cancel receiver")
        val intentFilter = IntentFilter()
        intentFilter.addAction(ACTION_DOWNLOAD_CANCEL)
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT)
        ContextCompat.registerReceiver(appContext, cancelReceiver, intentFilter, ContextCompat.RECEIVER_NOT_EXPORTED)

        // notifications
        Notifier.initChannels(appContext)
    }

    /**
     * onCreateView
     *
     * @param inflater inflater
     * @param container container
     * @param savedInstanceState saved instance state
     * @return
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(layoutId, container, false)
    }

    /**
     * onViewCreated
     *
     * @param view view
     * @param savedInstanceState saved instance state
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // components
        progressBar = view.findViewById(R.id.progressBar)
        progressStatus = view.findViewById(R.id.progressStatus)
        statusTextView = view.findViewById(R.id.status)

        // default button resources
        downloadButtonImageResId = R.drawable.bn_download
        downloadButtonBackgroundResId = R.drawable.bg_button_action

        // buttons
        downloadButton = view.findViewById(R.id.downloadButton)
        cancelButton = view.findViewById(R.id.cancelButton)
        deployButton = view.findViewById(R.id.deployButton)
        md5Button = view.findViewById(R.id.md5Button)

        // click listeners
        downloadButton.setOnClickListener {
            val wasNotDownloading = isDownloading.getAndSet(true)
            if (wasNotDownloading) {
                onDownloadStart()
                startUI()
                try {
                    start()
                } catch (e: NullPointerException) {
                    Log.e(TAG, "While starting", e)
                    warn(getString(R.string.status_download_null_exception))
                    onDownloadDone(Status.STATUS_FAILED, null)
                    isDownloading.set(false)
                } catch (e: Exception) {
                    Log.e(TAG, "While starting", e)
                    val message = e.message
                    warn(message ?: e.javaClass.name)
                    onDownloadDone(Status.STATUS_FAILED, null)
                    isDownloading.set(false)
                }
            }
        }
        cancelButton.setOnClickListener {
            cancelButton.visibility = View.INVISIBLE
            cancel()
            isDownloading.set(false)
        }
        deployButton?.setOnClickListener { deploy() }
        md5Button?.setOnClickListener { md5() }

        // source textviews
        setSource(view)

        // destination
        setDestination(view)

        // status
        statusTextView.text = ""

        // init
        if (savedInstanceState === null) {
            initUI()
        }
    }

    /**
     * Called when all saved state has been restored into the view hierarchy
     * of the fragment.  This can be used to do initialization based on saved
     * state that you are letting the view hierarchy track itself, such as
     * whether check box widgets are currently checked.  This is called
     * after [.onViewCreated] and before [.onStart].
     *
     * @param savedInstanceState If the fragment is being re-created from
     * a previous saved state, this is the state.
     */
    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        if (savedInstanceState != null) {

            downloadButton.visibility = savedInstanceState.getInt(DOWNLOAD_BTN_STATE, View.VISIBLE)
            downloadButtonImageResId = savedInstanceState.getInt(DOWNLOAD_BTN_IMAGE_STATE, R.drawable.bn_download)
            val drawable = AppCompatResources.getDrawable(requireContext(), downloadButtonImageResId)!!
            downloadButton.setImageDrawable(drawable)
            downloadButtonBackgroundResId = savedInstanceState.getInt(DOWNLOAD_BTN_BACKGROUND_STATE, R.drawable.bg_button_action)
            downloadButton.setBackgroundResource(downloadButtonBackgroundResId)

            progressBar.visibility = savedInstanceState.getInt(PROGRESS_STATE, View.INVISIBLE)
            progressStatus.visibility = savedInstanceState.getInt(PROGRESS_STATUS_STATE, View.INVISIBLE)

            cancelButton.visibility = savedInstanceState.getInt(CANCEL_BTN_STATE, View.INVISIBLE)
            deployButton?.visibility = savedInstanceState.getInt(DEPLOY_BTN_STATE, View.INVISIBLE)
            md5Button?.visibility = savedInstanceState.getInt(MD5_BTN_STATE, View.INVISIBLE)
        }
    }

    /**
     * onSaveInstanceState
     *
     * @param outState outgoing saved state
     */
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(DOWNLOAD_BTN_STATE, downloadButton.visibility)
        outState.putInt(DOWNLOAD_BTN_IMAGE_STATE, downloadButtonImageResId)
        outState.putInt(DOWNLOAD_BTN_BACKGROUND_STATE, downloadButtonBackgroundResId)
        outState.putInt(CANCEL_BTN_STATE, cancelButton.visibility)
        deployButton?.visibility?.let { outState.putInt(DEPLOY_BTN_STATE, it) }
        md5Button?.visibility?.let { outState.putInt(MD5_BTN_STATE, it) }
        outState.putInt(PROGRESS_STATE, progressBar.visibility)
        outState.putInt(PROGRESS_STATUS_STATE, progressStatus.visibility)
    }

    /**
     * onDestroy
     */
    override fun onDestroy() {
        super.onDestroy()

        // main receiver
        Log.d(TAG, "Unregister cancel receiver")
        appContext.unregisterReceiver(cancelReceiver)
    }

    // C A N C E L   R E C E I V E R

    /**
     * Broadcast receiver for cancel events
     */
    private val cancelReceiver: BroadcastReceiver = object : BroadcastReceiver() {

        override fun onReceive(context: Context, intent: Intent) {
            Log.d(TAG, "Received broadcast request")
            val action = intent.action
            if (ACTION_DOWNLOAD_CANCEL == action) {
                Log.d(TAG, "Received cancel broadcast request")

                // do
                cancel()
            }
        }
    }

    // O P E R A T I O N S

    /**
     * Start download
     */
    protected abstract fun start()

    /**
     * Cancel download
     */
    protected abstract fun cancel()

    /**
     * Deploy tail operation
     *
     * @noinspection EmptyMethod
     */
    protected abstract fun deploy()

    /**
     * Md5 operation
     */
    protected abstract fun md5()

    /**
     * Cleanup after download
     *
     * @noinspection EmptyMethod
     */
    @Suppress("EmptyMethod")
    protected abstract fun cleanup()

    // S T A T U S

    /**
     * Get reason
     *
     * @return reason
     */
    abstract val reason: String?

    /**
     * Build status message
     *
     * @param status status
     * @return message
     */
    private fun buildStatusString(status: Status?): String {
        val statusStr = status?.toString(appContext) ?: "unknown"
        if (status != Status.STATUS_SUCCEEDED) {
            if (reason != null) {
                return "$statusStr\n$reason"
            }
        }
        return statusStr
    }

    // U I   H E L P E R S

    /**
     * Initial UI state
     */
    @UiThread
    private fun initUI() {
        if (!isDownloading.get()) {
            downloadButton.visibility = View.VISIBLE
            progressBar.visibility = View.INVISIBLE
            progressStatus.visibility = View.INVISIBLE
            cancelButton.visibility = View.GONE
        } else {
            downloadButton.visibility = View.INVISIBLE
            progressBar.visibility = View.VISIBLE
            progressStatus.visibility = View.VISIBLE
            cancelButton.visibility = View.VISIBLE
        }
    }

    /**
     * Start UI state
     */
    @UiThread
    private fun startUI() {
        // buttons
        downloadButton.isEnabled = false
        downloadButton.visibility = View.INVISIBLE
        cancelButton.visibility = View.VISIBLE

        // progress
        progressBar.visibility = View.VISIBLE
        progressStatus.visibility = View.VISIBLE

        // status
        statusTextView.text = ""
    }

    /**
     * End UI state
     *
     * @param status status
     */
    @UiThread
    protected fun endUI(status: Status) {
        Log.d(TAG, "End UI $status")

        // progress
        progressBar.visibility = View.INVISIBLE
        progressStatus.visibility = View.INVISIBLE

        // status
        val message = buildStatusString(status)
        statusTextView.text = message
        statusTextView.visibility = View.VISIBLE

        // buttons
        downloadButtonImageResId = if (status == Status.STATUS_SUCCEEDED) R.drawable.bn_download_ok else R.drawable.bn_download
        val drawable = AppCompatResources.getDrawable(requireContext(), downloadButtonImageResId)!!
        downloadButton.setImageDrawable(drawable)
        downloadButtonBackgroundResId = if (status == Status.STATUS_SUCCEEDED) R.drawable.bg_button_ok else R.drawable.bg_button_err
        downloadButton.setBackgroundResource(downloadButtonBackgroundResId)
        downloadButton.isEnabled = false
        downloadButton.visibility = View.VISIBLE

        cancelButton.visibility = View.GONE
    }

    /**
     * Update UI
     */
    @UiThread
    protected fun updateUI() {

        val progress100 = if (progress!!.second == 0L) -1 else (progress!!.first * 100L / progress!!.second).toInt()
        val count = "$progress100%   ${formatAsInformationString(progress!!.first)}"
        val message = buildStatusString(status)
        Log.d(TAG, "Update UI $count, $message")
        progressBar.visibility = View.VISIBLE
        progressStatus.visibility = View.VISIBLE
        progressBar.progress = progress100
        progressStatus.text = count
        statusTextView.text = message
    }

    /**
     * Set source
     * @param view view
     */
    @UiThread
    private fun setSource(view: View) {
        val srcView = view.findViewById<TextView>(R.id.src)
        val srcView2 = view.findViewById<TextView>(R.id.src2)
        val srcView3 = view.findViewById<TextView>(R.id.src3)
        if (srcView2 != null && srcView3 != null) {
            try {
                val uri = downloadUrl?.toUri()
                var host = uri?.host
                val port = uri?.port
                if (port != -1) {
                    host += ":$port"
                }
                val file = uri?.lastPathSegment
                var path = uri?.path
                if (path != null && file != null) {
                    path = path.take(path.lastIndexOf(file))
                }
                srcView3.text = host
                srcView2.text = path
                srcView.text = file
            } catch (_: Exception) {
                srcView3.setText(R.string.status_error_invalid)
                srcView2.setText(R.string.status_error_invalid)
                srcView.setText(R.string.status_error_invalid)
            }
        } else {
            srcView.text = downloadUrl
        }
    }

    /**
     * Set destination
     * @param view view
     */
    @UiThread
    protected abstract fun setDestination(view: View)

    /**
     * Warn
     */
    private fun warn(message: CharSequence) {
        val activity: Activity? = activity
        if (activity != null && !isDetached && !activity.isFinishing && !activity.isDestroyed) {
            activity.runOnUiThread { Toast.makeText(activity, message, Toast.LENGTH_LONG).show() }
        }
    }

    // R E C O R D

    /**
     * Record target.
     * Target is either the downloaded file or the file contained in it and deployed from it.
     */
    protected fun recordTarget() {
        val arguments = arguments
        if (arguments != null) {
            val target = arguments.getString(DOWNLOAD_TARGET_FILE_ARG)
            if (target != null) {
                Settings.recordDatapackFile(appContext, File(target))
                Log.d(TAG, "Recorded $target")
            }
        }
    }

    // B R O A D C A S T

    /**
     * New data available request runnable, run when new data becomes available
     */
    var requestNew: Runnable? = null

    /**
     * Kill old data request runnable, run when old data should be discarded
     */
    var requestKill: Runnable? = null

    // E V E N T S

    /**
     * Event sink for download events fired by downloader
     */
    @Suppress("EmptyMethod")
    @CallSuper
    open fun onDownloadStart() {
    }

    /**
     * Event sink for download events fired by downloader
     *
     * @param status download status
     * @param downloadData download data
     */
    @CallSuper
    open fun onDownloadDone(status: Status, downloadData: DownloadData?) {
    }

    /**
     * Fire completion events to activity listener
     *
     * @param success whether download and deployment were successful
     */
    @CallSuper
    fun fireComplete(success: Boolean) {
        Log.d(TAG, "OnComplete succeeded=$success $this")
        val listener = (activity as CompletionListener?)!!
        listener.onComplete(success)
    }

    companion object {

        private const val TAG = "BaseDownloadF"
    }
}
